src/ol/renderer/webgl/TileLayerBase.js

```diff
@@ -0,0 +1,748 @@
+ */
+function getRenderExtent(frameState, extent) {
+  const layerState = frameState.layerStatesArray[frameState.layerIndex];
+  if (layerState.extent) {
+    extent = getIntersection(
+      extent,
+      fromUserExtent(layerState.extent, frameState.viewState.projection)
+    );
+  }
+  const source = /** @type {import("../../source/Tile.js").default} */ (
+    layerState.layer.getRenderSource()
+  );
+  if (!source.getWrapX()) {
+    const gridExtent = source
+      .getTileGridForProjection(frameState.viewState.projection)
+      .getExtent();
+    if (gridExtent) {
+      extent = getIntersection(extent, gridExtent);
+    }
+  }
+  return extent;
+}
+
+export function getCacheKey(source, tileCoord) {
+  return `${source.getKey()},${getTileCoordKey(tileCoord)}`;
+}
+
+/**
+ * @typedef {Object} Options
+ * @property {Object<string, import("../../webgl/Helper").UniformValue>} [uniforms] Additional uniforms
+ * made available to shaders.
+ * @property {number} [cacheSize=512] The tile representation cache size.
+ */
+
+/**
+ * @typedef {import("../../layer/BaseTile.js").default} BaseLayerType
+ */
+
+/**
+ * @classdesc
+ * Base WebGL renderer for tile layers.
+ * @template {BaseLayerType} LayerType
+ * @template {import("../../Tile.js").default} TileType
+ * @template {import("../../webgl/BaseTileRepresentation.js").default<TileType>} TileRepresentation
+ * @extends {WebGLLayerRenderer<LayerType>}
+ */
+class WebGLBaseTileLayerRenderer extends WebGLLayerRenderer {
+  /**
+   * @param {LayerType} tileLayer Tile layer.
+   * @param {Options} options Options.
+   */
+  constructor(tileLayer, options) {
+    super(tileLayer, {
+      uniforms: options.uniforms,
+    });
+
+    /**
+     * The last call to `renderFrame` was completed with all tiles loaded
+     * @type {boolean}
+     */
+    this.renderComplete = false;
+
+    /**
+     * This transform converts representation coordinates to screen coordinates.
+     * @type {import("../../transform.js").Transform}
+     * @private
+     */
+    this.tileTransform_ = createTransform();
+
+    /**
+     * @type {Array<number>}
+     * @protected
+     */
+    this.tempMat4_ = createMat4();
+
+    /**
+     * @type {import("../../TileRange.js").default}
+     * @private
+     */
+    this.tempTileRange_ = new TileRange(0, 0, 0, 0);
+
+    /**
+     * @type {import("../../tilecoord.js").TileCoord}
+     * @private
+     */
+    this.tempTileCoord_ = createTileCoord(0, 0, 0);
+
+    /**
+     * @type {import("../../size.js").Size}
+     * @private
+     */
+    this.tempSize_ = [0, 0];
+
+    const cacheSize = options.cacheSize !== undefined ? options.cacheSize : 512;
+    /**
+     * @type {import("../../structs/LRUCache.js").default<TileRepresentation>}
+     * @protected
+     */
+    this.tileRepresentationCache_ = new LRUCache(cacheSize);
+
+    /**
+     * @protected
+     * @type {import("../../Map.js").FrameState|null}
+     */
+    this.frameState_ = null;
```

Comment:
I think we are still consistently using the `_` suffix to signal that a property is private.  Could you rename these protected properties to conform with that?

(At some point we should add a linter rule about this if we want to keep the convention.)